### DISCUSSION TITLE
AVA 4

### DISCUSSION THREAD

**Discussion by novemberborn** [avajs/ava#2926] [General]
We're celebrating the new year with the official AVA 4 release! 🎊 

```console
npm install -D ava
```

### Worker threads 🧑‍💼

By default, test files are now run in worker threads! Huge thanks to @dnlup for landing this.

Test files should run a little quicker, since it's easier to spin up a worker thread than a child process. You can use `--no-worker-threads` on the command line, or `workerThreads: false` in your AVA configuration, to fall back to using child processes.

### Shared workers are no longer experimental 🎊

[Shared workers](https://github.com/avajs/ava/blob/main/docs/recipes/shared-workers.md) are no longer experimental. This is a powerful feature that loads a program in a worker thread in AVA's main process and then communicate with code running in the test workers. This enables your tests to better utilize shared resources during a test run, as well as providing opportunities to set up these resources before tests start (or clean them up after).

When you use watch mode, shared workers remain loaded across runs.

### Improved test selection 🤳

AVA selects test files based on your `package.json` or `ava.config.*` configuration files. When used on the CLI you can then provide patterns to select a subset of these files.

You can now pass a folder and AVA will select the test files contained therein. Or you can provide a specific JavaScript file and AVA will run it even if it wasn’t selected by your configuration.

If AVA finds files based on the configuration, but none of those were selected to the CLI arguments, it now prints a warning.

### Better monorepo support 🚝

AVA now looks for `ava.config.*` files in parent directories, until it finds a directory with a `.git` directory (or file). This lets you specify AVA configuration at the top of your monorepo and still run AVA from within each package.

### New snapshot format 📸

@ninevra has worked tirelessly on landing a new snapshot format. It contains all information necessary to regenerate the report file. This allows for snapshots to be updated even if tests or assertions are skipped.

### Previously failing test files run first 🏃

AVA now records which test files failed in the previous run. It then prioritizes testing these files when you run AVA again. Thanks @bunysae!

### ESM support 🤾

AVA 4 comes with full ES module support. Both ESM and CJS entrypoints are provided so that you can use it no matter how you manage your project.

The `ava.config.js` file is now treated as CJS or ESM depending on module type configured in the `package.json` file. `ava.config.mjs` is now supported.

If you use JavaScript files with non-standard extensions you can [configure AVA to import them](https://github.com/avajs/ava/blob/main/docs/06-configuration.md#configuring-module-formats).

Note that [dependency tracking in watch mode does not yet work](https://github.com/avajs/ava/issues/2388) with ES modules.

### Dedicated macro factory with type inference 🏭

`test.macro()` returns an object that can be used with `test()` and hooks. The `t.context` type is inherited from `test`. When used with TypeScript this gives much better type inference.

Like with AVA 3, regular functions that also have a `title` property that is a string-returning function are supported. However the type checking won’t be as good.

### Source maps 🗺

AVA now uses the source map support that’s built in to Node.js itself. This should give better stack traces. However we’re not sure yet what happens if you load a program that automatically modifies stack traces.

Line number selection (where `npx ava test.js:5` runs the test at line 5) now uses source maps and so should work better with TypeScript files.

### Assertions as type guards 💂

Most assertions now return a boolean indicating whether they passed. If you use AVA with TypeScript, you can use this as a type guard. Thanks @jmarkham828!

(This is not supported for `t.snapshot()` and the "throws" assertions.)

## Breaking changes 💥

AVA 4 requires at least Node.js 12.22, 14.17, 16.4 or 17. Node.js 10 is no longer supported.

If installed globally, AVA will no longer use any locally installed version. Install locally and run with `npx ava` instead. When running test files from another project that also has AVA installed, those test files will now fail to run (because they'll try and use that other AVA version).

### Ecosystem 🏞

* Support for `@ava/babel` has been removed. We haven’t seen enough [contributions to that package](https://github.com/avajs/babel) to recommend it for AVA 4. We’d be open to reinstating it in the future (just as we’d be open to support any other compilation provider).
* As a consequence, “enhanced assertions” are no longer available.
* AVA 4 requires `avajs/typescript@1.1` or newer.
* Support for the [`esm` package](https://www.npmjs.com/package/esm) has been removed.

### Configuration 🗒

* `ava.config.js` now follows the module type configured in `package.json`.
* `ava.config.*` files may be found that are outside your project directory.

### Tests and assertions 🥼

* By default test files execute in worker threads, not child processes.
* `test.meta.file` and `test.meta.snapshotDirectory` are now file URL strings.
* Whitespace in test titles is now normalized. This could result in two tests being treated as having the same title, even though they are different in code. Thanks @KillyMXI!
* `test()` and `t.try()` no longer take an array of test implementations. Use a loop instead.
* The `t.throws()` and `t.throwAsync()` assertions can no longer be called with a `null` value for the expectations argument.
* `test.cb()` and `t.end()` have been removed. Use async functions and [`util.promisify()`](https://nodejs.org/docs/latest/api/util.html#util_util_promisify_original) instead.
* `t.teardown()` now executes in last-in-first-out order.

### Snapshots 📸

* Snapshots recorded using earlier AVA versions are no longer recognized. Run `npx ava -u` to rebuild your snapshots after upgrading.
* Snapshots no longer recognize React elements, instead we want to provide this functionality [through a plugin interface](https://github.com/avajs/ava/pull/2428)
* `t.snapshot()` no longer works in hooks.
* `t.snapshot()` no longer takes an options argument allowing you to customize the snapshot ID.

### TypeScript 🪛

* The `instanceOf` expectation of `t.throws()` and `t.throwsAsync()` assertions must now be an `Error` constructor.
* You’re expected to use `test.macro()` when declaring macros.
* Implementation arguments now default to `unknown`.
* Types have been renamed. `Meta` replaces `MetaInterface`, other types with `Interface` suffixes now use the `Fn` suffix. There may be some other changes too. 4b4b2f63f184056b4038dfc692f99421038d3ec1
* `t.throws()` and `t.throwsAsync()` return `undefined` when the assertion fails. The type definition now matches this behavior.
* Our TypeScript definitions are now tested against TypeScript 4.4

## Other changes 🤓

* We’ve removed the non-verbose reporter because it proved too hard to maintain two subtly different reporter outputs. [We’d really like to rewrite the reporter.](https://github.com/avajs/ava/issues/2500)
* AVA's configuration files may now export promises or asynchronous factory methods.
* The `--config` argument may now point to a file that is not alongside the `package.json` file.
* When you use `t.timeout()`, AVA itself won’t time out until your test does. Thanks @OhYash!
* Multi-line snapshot labels now render correctly in Markdown. Thanks @KillyMXI!
* The reporters now clean up `.test` and `.spec` extensions from file names, as well as `test-` prefixes.
* Watch mode is better at detecting changes to snapshots and *not* rerunning tests.

## New Contributors

* @Fem-Fem made their first contribution in https://github.com/avajs/ava/pull/2750
* @OhYash made their first contribution in https://github.com/avajs/ava/pull/2758
* @KillyMXI made their first contribution in https://github.com/avajs/ava/pull/2773
* @mcecode made their first contribution in https://github.com/avajs/ava/pull/2794
* @duncanfinney made their first contribution in https://github.com/avajs/ava/pull/2820
* @asaid-0 made their first contribution in https://github.com/avajs/ava/pull/2788
* @bompus made their first contribution in https://github.com/avajs/ava/pull/2843
* @StefanoA1 made their first contribution in https://github.com/avajs/ava/pull/2849
* @live627 made their first contribution in https://github.com/avajs/ava/pull/2851
* @alessandroasm made their first contribution in https://github.com/avajs/ava/pull/2863
* @eltociear made their first contribution in https://github.com/avajs/ava/pull/2870
* @hamidbae made their first contribution in https://github.com/avajs/ava/pull/2878
* @rrichardson made their first contribution in https://github.com/avajs/ava/pull/2910
* @make-github-pseudonymous-again made their first contribution in https://github.com/avajs/ava/pull/2913
* @lightmare made their first contribution in https://github.com/avajs/ava/pull/2918

**Full changelog since RC 1**: https://github.com/avajs/ava/compare/v4.0.0-rc.1...v4.0.0
**Full changelog since AVA 3**: https://github.com/avajs/ava/compare/v3.15.0...v4.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/avajs/ava/releases/tag/v4.0.0'>AVA 4</a>.</em>

**Comment by Brooooooklyn** [comment]
`wokerThreads: true` cause `0xc0000005 (Access Violation)` on Windows. See https://github.com/napi-rs/napi-rs/runs/4697819504?check_suite_focus=true

Disabled `workerThreads` could fix it: https://github.com/napi-rs/napi-rs/pull/1004

**Reply by novemberborn** [reply]
I think this is a limitation in Node-API or perhaps something your project needs to deal with. I suspect your code won't work in worker threads at all. This shouldn't be AVA specific.

**Comment by tommy-mitchell** [comment]
What was the reasoning behind the change to `t.throws` and `t.throwsAsync` to possibly return `undefined`?

**Reply by novemberborn** [reply]
That was just a type change to match the runtime behavior. The assertion returns `undefined` if it did not capture an error.

**Reply by tommy-mitchell** [reply]
Do tests not just end if it doesn’t throw?

**Reply by novemberborn** [reply]
AVA's assertions don't throw when they fail, so no the tests keep running.
